_posts/2018-10-09-enabling-ecc-on-pc-engines-apu2.md

```diff
@@ -0,0 +1,354 @@
+---
+post_title: Enabling ECC on PC Engines platforms
+author: Krystian Hebel
+layout: post
+published: true
+post_date: 2018-10-09 15:00:00
+
+tags:
+    - ECC
+    - APU2
+categories:
+    - Firmware
+
+---
+
+## Introduction
+
+Sometimes a bit in RAM changes its value spontaneously due to electrical or
+magnetic interference. It can be caused by background radiation, cosmic rays or
+recently [attacks](https://googleprojectzero.blogspot.com/2015/03/exploiting-dram-rowhammer-bug-to-gain.html)
+using [row hammering](https://en.wikipedia.org/wiki/Row_hammer).
+Error-correcting code (ECC) memory helps with mitigation of this problem by
+adding more data storage for storing information that makes detection and
+correction of errors possible. Memory controller scans whole ECC enabled memory,
+reading every piece of data, checking (and correcting if necessary) for errors
+and writing data back to memory. This process is called memory scrubbing. As ECC
+can correct only limited number of flipped bits scrubbing has to be done
+periodically, before multiple errors within one ECC word can occur - time
+between scrubs isn't fixed, and according to
+[BKDG, 52740 Rev 3.06](https://www.amd.com/system/files/TechDocs/52740_16h_Models_30h-3Fh_BKDG.pdf):
+
+> There are many factors which influence scrub rates. Among these are:
+> * The size of memory or cache to be scrubbed
+> * Resistance to upsets
+> * Geographic location and altitude
+> * Alpha particle contribution of packaging
+> * Performance sensitivity
+> * Risk aversion
+
+Usually, ECC can fix only single bit, and detect two changed bits in ECC word,
+but with this SoC a variation named `x4 ECC` is used:
```

Comment:
@krystian-hebel let's not use `this SoC` but fully qualified name `AMD Embedded G series GX-412TC SoC included in PC Engines apu2 platform` or something like that.